src/main/java/uk/gov/justice/digital/hmpps/cmd/api/service/NotificationService.kt

```diff
@@ -37,68 +37,86 @@ class NotificationService(
         val csrClient: CsrClient
 ) {
 
-    fun getNotifications(processOnReadParam: Optional<Boolean>, unprocessedOnlyParam: Optional<Boolean>, fromParam: Optional<LocalDate>, toParam: Optional<LocalDate>): Collection<NotificationDto> {
+    fun getNotifications(processOnReadParam: Optional<Boolean>, unprocessedOnlyParam: Optional<Boolean>, fromParam: Optional<LocalDate>, toParam: Optional<LocalDate>, quantumId: String = authenticationFacade.currentUsername): Collection<NotificationDto> {
         val start = calculateStartDateTime(fromParam, toParam)
         val end = calculateEndDateTime(toParam, start)
         return getShiftNotificationDtos(start, end, unprocessedOnlyParam.orElse(false), processOnReadParam.orElse(true))
     }
 
     fun sendNotifications() {
-        val unprocessedNotifications = shiftNotificationRepository.findAllByProcessedIsFalse()
-        log.info("Sending notifications, found: ${unprocessedNotifications.size}")
-
-        unprocessedNotifications.groupBy { it.quantumId.toUpperCase() }
+        log.info("Sending Notifications")
+        shiftNotificationRepository.findAllByProcessedIsFalse()
+                .groupBy { it.quantumId }
                 .forEach { group ->
-                    try {
+                    try{
                         sendNotification(group.key, group.value)
                         group.value.forEach { it.processed = true }
                     } catch (e: NotificationClientException) {
                         log.warn("Sending notifications to user ${group.key} FAILED", e)
                     }
-                    log.info("Sent notification (${group.value.size} lines) for ${group.key}")
                 }
         log.info("Finished sending notifications")
     }
 
-    fun generateAndSaveNotifications() {
+    fun refreshNotifications() {
+        log.info("Refreshing notifications")
         val allPrisons = prisonService.getAllPrisons().distinctBy { it.csrPlanUnit }
-        val newShiftNotifications = allPrisons
-                .flatMap { prison ->
-                    csrClient.getShiftNotifications(prison.csrPlanUnit, prison.region)
-                }.distinct()
-                .map{
-                    if(it.actionType == ShiftActionType.EDIT.value && !checkIfEditNotificationsHasCorrespondingAdd(it.quantumId, it.shiftDate, it.shiftType, it.shiftModified)) {
-                        it.actionType = ShiftActionType.ADD.value
-                    }
-                    it
+        log.info("Found ${allPrisons.size} prisons")
+        allPrisons
+                .forEach { prison ->
+                    saveNotifications(
+                        csrClient.getShiftNotifications(prison.csrPlanUnit, prison.region)
```

Comment:
Save all for each prisons results rather than at the end, should make the notifications feel more real time without actually running more often.